File: Linked List/Easy/160.Intersection of Two Linked Lists/Intersection of Two Linked Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        ListNode *curA, *curB;
        curA = headA;
        curB = headB;
        if(curA == NULL || curB == NULL){
            return NULL;
        }
        int length_a = getLength(curA);
        int length_b = getLength(curB);
        if(length_a > length_b){
            for(int i=0; i < length_a-length_b; i++){
                curA = curA->next;
            }
        }
        else if(length_a < length_b){
            for(int i=0; i < length_b-length_a; i++){
                curB = curB->next;
            }
        }
        while(curA != curB){
            curA = curA->next;
            curB = curB->next;
        }
        return curA;
    }
private:
    int getLength(ListNode *head){
        int length = 1;
        while(head->next != NULL){
            head = head->next;
            length++;
        }
        return length;
    }
};

File: Linked List/Easy/160.Intersection of Two Linked Lists/Intersection of Two Linked Lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        if headA is None or headB is None:
            return None

        pa = headA # 2 pointers
        pb = headB

        while pa is not pb:
            # pa先遍历headA，然后再遍历headB
            # pb先遍历headB，然后再遍历headA
            pa = headB if pa is None else pa.next
            pb = headA if pb is None else pb.next

        return pa # 只有两种方式结束循环，一种是pa和pb所指相同，另一种是headA和headB都已经遍历完仍然没有找到。
